modin/pandas/indexing.py

```diff
@@ -696,11 +696,18 @@ def __setitem__(self, key, item):
         pandas.DataFrame.loc
         """
         row_loc, col_loc, _ = self._parse_row_and_column_locators(key)
-        if isinstance(row_loc, list) and len(row_loc) == 1:
-            if row_loc[0] not in self.qc.index:
-                index = self.qc.index.insert(len(self.qc.index), row_loc[0])
-                self.qc = self.qc.reindex(labels=index, axis=0)
-                self.df._update_inplace(new_query_compiler=self.qc)
+        if (
+            isinstance(row_loc, list)
+            and len(row_loc) == 1
+            and row_loc[0] not in self.qc.index
+        ) or (
+            not (is_list_like(row_loc) or isinstance(row_loc, slice))
+            and row_loc not in self.qc.index
+        ):
+            row_loc = row_loc[0] if isinstance(row_loc, list) else row_loc
+            index = self.qc.index.insert(len(self.qc.index), row_loc)
+            self.qc = self.qc.reindex(labels=index, axis=0)
+            self.df._update_inplace(new_query_compiler=self.qc)
```

Comment:
```suggestion
       self._maybe_compute_enlarge_labels(row_loc, col_loc)
```
I guess we actually have a method in the Locator class that was brought to solve exactly the same problem but is unused for some reason: [`_compute_enlarge_labels`](https://github.com/modin-project/modin/blob/d590de080dddb7ed715a58d57c602d96322c4dcf/modin/pandas/indexing.py#L725).

We may try to replace the current `_compute_enlarge_labels` logic with the introduced one and then replace this if-statement with a single call of `self._maybe_compute_enlarge_labels`.

Reply:
Wouldn't that require me to compute a new index with the labels from the loc and then call this function? Because the input to the function you linked takes two Indexes and computes the differences, and in this case we already have the difference?

Also should I hold off on merging and change this, or should this be another PR?

Reply:
The initial idea of my suggestion was to hide this scary if-construction into some function. We already have a function that is supposed to compute enlargest labels, but it seems that it doesn't do exactly what we want. As far as I understand, it mimics some outdated pandas logic and isn't used anywhere in the code. That's why I propose to remove `_compute_enlarge_labels` at all and replace it with some `_maybe_enlarge_labels` function that would take row and column locators and perform the same logic that this PR is introducing (check whether `row_loc` contains a label that's not presented in the index and insert the missing label into the index if needed).

I would like `_maybe_enlarge_labels` to be added in this PR, as for the removal of `_compute_enlarge_labels` - I think we should do it in a separate PR.

Reply:
That makes sense to me! From a naming perspective do we want to name it something like `_enlarge_labels`? Happy either way though!

Reply:
Also this special case only applies if we are indexing a single value that doesn't exist. Should this be the case, or should we also be checking if lists with multiple indexers have missing labels and adding them?
e.g. if a df has a range index from 0 to 2, and we do df[3, 4, 5] = df[1] should we add index elements 3, 4, and 5? Because this would error in the current implementation @devin-petersohn @dchigarev

Reply:
what pandas does in this case? I guess it's what we have to imitate anyway

Reply:
So I ran some tests:
Pandas:
```python
>>> import pandas as pd
>>> df = pd.DataFrame([[1, 2], [3, 4]], columns=['A', 'B'])
>>> df.loc[2, 3] = df.loc[1]
ValueError: Incompatible indexer with Series
>>> df
     A    B   3
0  1.0  2.0 NaN
1  3.0  4.0 NaN
2  NaN  NaN NaN
# Reset df
>>> df.loc[[2, 3]] = df.loc[1]
KeyError: "None of [Int64Index([2, 3], dtype='int64')] are in the [index]"
>>> df.loc[[2, 3], :] = df.loc[1]
KeyError: "None of [Int64Index([2, 3], dtype='int64')] are in the [index]"
>>> df.loc[:, 'C', 'D'] = df.loc[:, 'A']
TypeError: unhashable type: 'slice'
>>> df.loc[:, ['C', 'D']] = df.loc[:, 'A']
>>> df
   A  B  C  D
0  1  2  1  3
1  3  4  1  3
```
Modin (without this fix):
```python
>>> import modin.pandas as pd
>>> df = pd.DataFrame([[1, 2], [3, 4]], columns=['A', 'B'])
>>> df.loc[2, 3] = df.loc[1]
KeyError: array([2])
>>> df
    A  B
0  1  2
1  3  4
>>> df.loc[[2, 3]] = df.loc[1]
KeyError: array([2, 3])
>>> df.loc[[2, 3], :] = df.loc[1]
KeyError: array([2, 3])
>>> df.loc[:, 'C', 'D'] = df.loc[:, 'A']
IndexingError: Too many indexers
>>> df.loc[:, ['C', 'D']] = df.loc[:, 'A']
KeyError: array(['C', 'D'], dtype=object)
>>> df
   A  B
0  1  2
1  3  4
```
so there are a lot of places where we are differing from `pandas` or where we also error but have different (less transparent) errors. 
Modin (with this fix so far):
```python
>>> import modin.pandas as pd
>>> df = pd.DataFrame([[1, 2], [3, 4]], columns=['A', 'B'])
>>> df.loc[2, 3] = df.loc[1]
KeyError: array([3])
>>> df
     A    B
0  1.0  2.0
1  3.0  4.0
2  NaN  NaN
# Reset df
>>> df.loc[[2, 3]] = df.loc[1]
KeyError: array([2, 3])
>>> df.loc[[2, 3], :] = df.loc[1]
KeyError: array([2, 3])
>>> df.loc[:, 'C', 'D'] = df.loc[:, 'A']
IndexingError: Too many indexers
>>> df.loc[:, ['C', 'D']] = df.loc[:, 'A']
KeyError: array(['C', 'D'], dtype=object)
>>> df
   A  B
0  1  2
1  3  4
```
I'm leaving this here to document the differences we have so we can start fixing them!